src/filter/or.rs

```diff
@@ -58,50 +60,37 @@ where
-                Ok(Async::Ready(ex1)) => {
-                    return Ok(Async::Ready((Either::A(ex1),)));
-                }
-                Ok(Async::NotReady) => return Ok(Async::NotReady),
-                Err(e) => e,
-            },
-            State::Second(ref mut err1, ref mut second) => {
-                return match second.poll() {
-                    Ok(Async::Ready(ex2)) => Ok(Async::Ready((Either::B(ex2),))),
-                    Ok(Async::NotReady) => Ok(Async::NotReady),
+    type Output = Result<(Either<T::Extract, U::Extract>,), <U::Error as CombineRejection<T::Error>>::Rejection>;
 
-                    Err(e) => {
-                        self.original_path_index.reset_path();
-                        let err1 = err1.take().expect("polled after complete");
-                        Err(e.combine(err1))
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let pin = get_unchecked!(self);
+        loop {
+            let (err1, fut2) = match pin.state {
+                State::First(ref mut first, ref mut second) => match ready!(pin_unchecked!(first).try_poll(cx)) {
+                    Ok(ex1) => {
+                        return Poll::Ready(Ok((Either::A(ex1),)));
                     }
-                };
-            }
-            State::Done => panic!("polled after complete"),
-        };
-
-        self.original_path_index.reset_path();
-
-        let mut second = match mem::replace(&mut self.state, State::Done) {
-            State::First(_, second) => second.filter(),
-            _ => unreachable!(),
-        };
+                    Err(e) => (e, second.filter()),
+                },
+                State::Second(ref mut err1, ref mut second) => {
+                    let ex2 = match ready!(pin_unchecked!(second).try_poll(cx)) {
+                        Ok(ex2) => {
+                            Ok((Either::B(ex2),))
+                        },
+                        Err(e) => {
+                            pin.original_path_index.reset_path();
+                            let err1 = err1.take().expect("polled after complete");
+                            Err(e.combine(err1))
+                        }
+                    };
+                    pin.state = State::Done;
+                    return Poll::Ready(ex2)
+                }
+                State::Done => panic!("polled after complete"),
+            };
 
-        match second.poll() {
-            Ok(Async::Ready(ex2)) => Ok(Async::Ready((Either::B(ex2),))),
-            Ok(Async::NotReady) => {
-                self.state = State::Second(Some(err1), second);
-                Ok(Async::NotReady)
-            }
-            Err(e) => {
-                self.original_path_index.reset_path();
-                return Err(e.combine(err1));
-            }
+            pin.original_path_index.reset_path();
+            pin.state = State::Second(Some(err1), fut2);
```

Comment:
This loop seems to be breaking `tests/path.rs#or`

Reply:
yeah, that one is solved thanks
some more to go 😀